### ISSUE TITLE
New TSConfig Management

### ISSUE THREAD

**Issue by noelmace** [dherges/ng-packagr#797]
## Type of Issue

```
[ ] Bug Report
[ ] Feature Request
[x] discussion
```

## Description

This discussion is created in order to permit a more global discussion about TSConfig management, customisation and refactoring, directly following https://github.com/dherges/ng-packagr/pull/790#discussion_r183844424.

**Comment by noelmace** [comment]
## current mechanism

***required -> consumer -> override (see #786)***

### required

Was created in order to keep the "right" error message, that is to say the one from TSC, which, in my opinion, reflect the best the reason of the error (at least, better than a "your TSConfig doesn't respect the requirements of ng-packagr").

If we want a better error message, we'll need to write one for each case. 

### override

are options that will break the ng-packagr build entirely (without a readable error, as we could see with all the related issues which was really difficult to trace).

### Why

1. writing a tsconfig could **always** break your build. TSC and NGC really have a better maturity (so, error messages) that ng-packagr on this for much cases
2. but ng-packagr always relies on some of this options in a really specific way, what leads to a set of mandatory options which will have to be overriden in order to hide some misleading erros

### How

The identification of the options and their sorting out between a and b cases, was proof tested  during https://github.com/angular/devkit/pull/750 ...

## Some other (futur) options

*(see options 1 and 2 proposed in #790 (comment) as this is a response to it)*

I personally don't think the overriding is a good long term solution, as it could mislead the user on the behaviour of ng-packagr. This would require some documentation at least (but doc is not fix :wink: ). I consider that this is a temporary solution, waiting for a better one. So, I disagree with the option 1. The option 2, at least, could be a better solution, but it relies again on documentation, and we can be sure that this will lead to misunderstandings.

All this question reflect the global approach of ng-packagr : wrapping a lot of tools instead of making just a new one.

Globally, there is some issues with the 3 ways tsconfig are generated and used, as it's misleading a lot of users when using a custom TSConfig (especially with imports / paths, like https://github.com/dherges/ng-packagr/issues/285#issuecomment-384101460, https://github.com/dherges/ng-packagr/issues/519#issuecomment-383418228 or https://github.com/dherges/ng-packagr/issues/791)

Following that, I personally see two other options, but this would need some more refactoring : 
1. simply hide the TSConfig as well (as it could be misleading), and only permit specific options
    a. by writing a new json schema
    b. or by adding a new typescript options object to ng-package (and removing the withTsConfig method from the public api, as advanced users could always use providers)
2. forget about any tsconfig overriding, and entirely rethink error handling on tsc, ngc, etc ... by wrapping them, as angular-cli tend to do for some cases. I don't think this could be a good solution for ng-packagr (at it would be too much work to refactor and then maintain), but I prefer to put it here just in case :slightly_smiling_face:

**Comment by alan-agius4** [comment]
I am not sure about the approaches, let me explain a bit further and put also some ideas on the table as well.

Replacing the option to provide a tsconfig file with providing a limited list of options is quite limiting in some cases as for instance myself, I use several options, `baseUrl`, `types`, `paths`, `disableTypeScriptVersionCheck`, `preserveConstEnums`, `strictPropertyInitialization` etc... And having to set these in 2 places. ie a `tsconfig` file and an `ng-packagr` config is less than ideal as it will increase maintainability and the risk of forgetting to add a propery in `ng-packagr` config. Setting these options in tsconfig is needed for the IDE to function/properly. 

When you read a tsconfig file, using `tsc` or `ngc` it does more than just merging the options, it resolves the paths and some other things. In my case I highly depend on the fact that I am using my `baseUrl` for these resolutions and not what `ng-packagr` requires for the build.

Having to set a custom JSON schema of all the properties in `tsc`, `ngc` is not scalable as with every `ts` and `ngc` version there are new properties added in the `compilerOptions` and we are talking at least 80 different options for the moment.

My two cents in order to avoid issues for consumers that use a custom tsconfig would be the following;

1. Better error messages when one of the mandatory option is missing or set inappropriately. example when setting `declaration: false` or it's not specified.
2. Create a schematic to generate a `tsconfig` with the mandatory options, for users that want to have there own tsconfig.
3. Document that fact that users can extend the `tsconfig` with the default `tsconfig` that `ng-packagr` provides.

**Comment by noelmace** [comment]
About the maintainability : the compiler-cli and ts api already defines the right typings and schemas, on which we could relie (https://github.com/angular/angular/blob/c8a1a14b87e5907458e8e87021e47f9796cb3257/packages/compiler-cli/src/perform_compile.ts#L104). I agree that this could lead to some more maintenance on this, but not as much as it would seems at first. Also, this could improve issues readability, as we would have a total control on the underlying tools configurations.

Even if I totally agree that this would have a big downside on IDE integration ... as a matter of fact, it is for now in most cases impossible to set a unique tsconfig for development environment (IDE etc ...) and ng-packagr in an angular-cli v6 mono-repository (/ multi-projects), as some options aren't used properly (for example, paths).

This is why I said that permitting to set a custom tsconfig.json file is missleading : it suppose that this tsconfig would be simply used by ng-packagr in the same way tsc and ngc does (as you said), and passed to the underlying tools without any modifications.

I don't see any option to do that, so, here is a little variation of what I proposed early, based on your feedback : **we could separate the logics**. Permitting to set a custom tsconfig for this kind of use cases, *and* add build options to ng-package config.

My feedbacks on your "two cents" : 
1. this is exactly why I didn't override this parameters, as I said early : the underlying tools give adequate error messages, so hadding our own would just decrease the maintainability without adding anything ... for now :wink: 
2. a big :+1: on this
3. :+1: again ... as what we are speaking about here would lead to an important refactoring, we should at least document the existing behaviour

I'm a bit overloaded with work for now, but I could start to do something about that in 2 weeks.

**Comment by alan-agius4** [comment]
If you take into consideration a mono repo for example, that you mentioned. It is vital that the tsconfig is not part of static options. And you have a `tsconfig` with a custom `baseUrl` and `paths` to work. In fact I do have a starter kit for this which you can see here.

https://github.com/alan-agius4/ng-mono-repo-starter/blob/master/tsconfig.json
https://github.com/alan-agius4/ng-mono-repo-starter/blob/master/tsconfig.build.json

Having full control of a tsconfig, Outside of ng-packagr is essential for this, As i rely on the ngc to properly parse the paths etc and not `ng-packagr`. As I said early, you cannot simply get some options and merge them with tsconfig, as this simple won't work because when a tsconfig is parsed, it does much more than merging an object.

My point is, we should allow a full custom `tsconfig`, this is something quite important in my use cases. In case the user provides invalid configuration than we should warn them. By removing the functionality would have a huge impact.

Just a small not on my point 3, you mentioned this requires refactoring, but in  reality it doesn't it just needs to revert this https://github.com/dherges/ng-packagr/pull/786/files, and it will just work out of the box.

**Comment by noelmace** [comment]
@alan-agius4  So, if I understand, you propose to reverte #786 because you think documentation + schematics would be enough, and that we can deal we the related issues in the meantime ? As it fix a lot of issues, doesn't introduce new ones, is consistent with the previous behaviour, was validate by @dherges who did this previous behaviour, and doesn't forbid a deeper refactoring (wich we are talking about here), I don't see why. Maybe @dherges could arbitrate here.

As, in your starter kit, tsconfig.build.json is used only by ng-packagr, I don't understand why a full custom tsconfig is so important in your use case. Could you please explain ? Talking about the tsconfig used outside of ng-packagr doesn't seems relevant in our case here. FYI, I'm, for my part, currently focusing on angular-cli v6 mono-repositories.

It seems to me that ng-packagr control entirely the resulting tsconfig which would be passed to the underlying tools *before* passing it, so, again, why could we not "simply get some options and merge them with tsconfig" ? This is exactly what is done is ng-packagr, even before #786 : https://github.com/noelmace/ng-packagr/blob/4e6c4f4bb21869c69f084560af3fa75531e4d07f/src/lib/ts/tsconfig.ts#L49

**Comment by noelmace** [comment]
I think we are talking about a lot of unrelated things here, and that we should focus on : 
1. the current behaviour and the possible fix on it
2. how we could enhance it in order to have better error handling and configuration (the purpose of this issue)

So, f I try to formalise what we talked about here : 

## first step : enhance the current behaviour

* [ ] document the current tsconfig requirements of ng-packagr :heavy_check_mark:

* [ ] keep #786, as a temporary solution VS create a schematic as an alternative :question: @dherges 
 
Following https://github.com/angular/devkit/pull/750#issuecomment-383127935, I think @filipesilva could have his opinion on this.

In my opinion, #786 is better as it wouldn't require any additional tool and is entirely transparent for the user. Frankly speaking, @alan-agius4, I neither see exactly what you think is wrong with thi, as ng-packagr is still at an early stage, nor why we should stop fixing simple issues like #790 on it, waiting for an hypothetical refactoring.

## second step : tsconfig management refactoring

* [ ] refactor in order to have a unique, central starting point for all tsconfig, with transformations on it for specific behaviours (https://github.com/angular/devkit/pull/750#issuecomment-383319671) :question: 
* [ ] remove all the required / override behaviour on the custom tsconfig which was passed to ng-packagr and use it directly has a real override of the default tsconfig file, but only as an advanced / "know what you're doing" feature :question: this could respond to the "extend" problematic
* [ ] add some simple build options in order to permit to customize the build in a much more simple way :question:

The final process would then be : 
```
default tsconfig OR user override
--> ng-packagr (configurable) transformations
--> tsconfig passed to the underlying tools
```

**Comment by alan-agius4** [comment]
@noelmace I am not against the fix per see. 

But I am questioning if that fix or feature is needed in the first place. Ie: https://github.com/dherges/ng-packagr/pull/786/files, and https://github.com/noelmace/ng-packagr/blob/4e6c4f4bb21869c69f084560af3fa75531e4d07f/src/lib/ts/tsconfig.ts#L49

There are two things here, when you want to create a `tsconfig`, you must create a valid `tsconfig`, if you create a tsconfig which is not valid example without `moduleResolution` even without `ng-packagr` for instance `ngc` build would fail. Because it is imperative, when you create a tsconfig, you must create a valid one. For example, if I have an ng app project with angular cli I cannot simple create a `tsconfig` without the required properties and expect it to work. (There is a schematic which helps in creating this for you). 

I do think the missing pieces here is documentation and a schematic. At the moment if you want to use the API and use a custom tsconfig, there are 2 ways, create a full valid tsconfig from scratch. extend the base `ng-packagr` tsconfig that is shipped with the library, and change whichever property you want to change.

For users, it might be confusing, if they are setting something in tsconfig, and it has no effect, because it is being overridden in code, while if you use the `extend`, you know that most likely it is because of that.  And mind you, what I am saying here, can be done now because it's already supported, no hypothetical refactoring required.

**Comment by alan-agius4** [comment]
Ps: I just had a look at the angular schematic, and now I understand that the tsconfig cannot be extended as it's already extended by the main one.

**Comment by noelmace** [comment]
Ok, I totally agree on that. The only difference in our two points of view seems then to be on the planning.

I did #786 (and then #790) in order to respond to https://github.com/angular/devkit/pull/750#issuecomment-383127935 and to fix some issues (for example, #285) quickly, following what was already in place.

I totally agree that all the required + override way is confusing (as https://github.com/dherges/ng-packagr/issues/519#issuecomment-384375281 show it, for example), but I think that we need quickly a solution for that.

Then, maybe we could settle on the two steps I proposed.

First, keep the present behaviour (not because I did #786, or I think it could be the right way in long term, but because I think it is a satisfying short term solution), then document all the requirements and, finally, update the schematics in devkit accordingly, for minor versions (without introducing any breaking change).

And then, as a second and final step, aiming at the next major version, refact all the build configuration (following, or not, the final process I proposed, as we have time to discuss that with the other contributors, mainly @dherges and @filipesilva), with the liberty to introduce breaking changes if needed.
